### PR TITLE
refactor zypper module

### LINKED ISSUE #299: Zypper always reports changed
Issue Type:

Bug Report

Ansible Version:

1.8.2

Environment:

RHEL 7

Summary:

The zypper module installs packages by running zypper in a shell.  It then checks the return code to determine if anything was changed.  However, zypper will return success (0) even when it reports "Nothing to do."

Steps To Reproduce:
1. Install a package with zypper

```
  ansible -m zypper -a 'name=ssh state=present' -i inventory sles_host
```
1. Do it again

```
  ansible -m zypper -a 'name=ssh state=present' -i inventory sles_host
```

Expected Results:

Ansible may report that a change occurred after the first step.
Ansible will report that no change occurred after the second step.

Actual Results:

Ansible reports that a change occurred after both steps.

### LINKED ISSUE #546: zypper feature request: name=* state=latest
To replicate the yum module functionality please allow for updating all packages.  Arguments to replicate this functionality would be "name=\* state=latest".   The command result would be "zypper update"

### LINKED ISSUE #1627: Ansible 2.0 zypper module type=patch state=present, always results in changed
`- zypper: type=patch name=openSUSE-2016-128 state=present 
`

result is always,

```
TASK [leap : zypper] ***********************************************************
changed: [10.192.165.194] => {"changed": true, "name": ["openSUSE-2016-128"], "rc": 0, "state": "present"}
```

### LINKED ISSUE #2094: zypper: does not install package from URL
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

zypper
##### ANSIBLE VERSION

```
ansible 2.0.1.0
```
##### SUMMARY

Zypper command is not installing package from URL
##### STEPS TO REPRODUCE

```
ansible all -i 'localhost,' -c local -m zypper -a "name='https://download.elasticsearch.org/elasticsearch/release/org/elasticsearch/distribution/rpm/elasticsearch/2.1.1/elasticsearch-2.1.1.rpm'"
```
##### EXPECTED RESULTS

RPM is installed via zypper.
##### ACTUAL RESULTS

Error comes back and says "No provider of 'elasticsearch' found."

### PR DESCRIPTION
**Pull request by robinro**
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

packaging/os/zypper.py
##### ANSIBLE VERSION

devel, 2.1.0
##### SUMMARY

This is a major cleanup/refactor of the zypper module.
Thanks to @alxgu, @agx, @jasonmader, @andytom, whose commits/PRs are included here.

Cleanup:
- remove mention of old_zypper (no longer supported)
  - requirement goes up to zypper 1.0, SLES 11.0, openSUSE 11.1
  - allows to use newer features (xml output)
  - already done for zypper_repository
- use zypper instead of rpm to get old version information, based on work by @jasonmader
- don't use rpm, zypper can do everything itself
- run zypper only twice, first to determine current state, then to apply changes

New features:
- determine change by parsing zypper xmlout
- determine failure by checking return code
- allow simulataneous installation/removal of packages (using '-' and '+' prefix)
  - allows to swap out alternatives without removing packages depending
    on them
- implement checkmode, using zypper --dry-run
- implement diffmode
- implement 'name=\* state=latest' and 'name=\* state=latest type=patch'
- add force parameter, handed to zypper to allow downgrade or change of vendor/architecture

Fixes/Replaces:
- fixes #1627, give changed=False on installed patches
- fixes #2094, handling URLs for packages
- fixes #1461, fixes #546, allow state=latest name='*'
- fixes #299, changed=False on second install, actually this was fixed earlier, but it is explicitly tested now
- fixes #1824, add type=application
- fixes #1256, install rpm from path, this is done by passing URLs and paths directly to zypper
- fixes #1262, made unnecessary by using zypper instead of rpm to get the current state

### CODE REVIEW

**Comment by robinro** [comment]
@alxgu, @agx, @jasonmader, @andytom pleave have a look and let me know what you think.

So far I only tested on openSUSE 42.1, but will also try/fix older openSUSE versions.

**Comment by robinro** [comment]
This is kind of a follow-up to the `zypper_repository` refactor in https://github.com/ansible/ansible-modules-extras/pull/1990
Also there `old_zypper` was deprecated and zyppers xml output used for everything.
@matze, @paavoap Maybe you can also test/comment?

**Comment by gregdek** [comment]
Thanks @robinro for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

File: packaging/os/zypper.py [hunk 1]
```diff
@@ -96,168 +108,236 @@
...
+    # cmd.extend(name_install)
+    # retvals['cmd'] = cmd
+    # result, retvals['rc'], retvals['stdout'], retvals['stderr'] = parse_zypper_xml(m, cmd)
+
+    # # if we've already made a change, we don't have to check whether a version changed
+    # if result:
+    #     # at least one package was updated
+    #     retvals['changed'] = True
+    #     if m._diff:
+    #         set_diff(m, retvals, result)
```
Comment (matze) [inline packaging/os/zypper.py]: Minor nitpick: the out-commented code could be removed.

**Comment by robinro** [comment]
Thanks @matze, I overlooked that part.
Also the version_added is now increased.

**Comment by andytom** [comment]
Looks good to me. The versions of openSuse and SLES that required `old_zypper` are no longer supported by Novell so I guess it is safe to drop support for them here.

**Comment by jasonmader** [comment]
@robinro I just had a chance to test this and had an issue when a repo signature verification failed:

```
"failed": true, "msg": "Zypper run command failed with return code 4.", "rc": 4
```

full text:

```
fatal: [grandizer.local]: FAILED! => {"changed": false, "cmd": ["/usr/bin/zypper", "--quiet", "--non-interactive", "--xmlout", "search", "--type", "patch", "--match-exact", "--verbose", "--installed-only", "openSUSE-2016-38", "openSUSE-2016-106"], "failed": true, "msg": "Zypper run command failed with return code 4.", "rc": 4, "stderr": "", "stdout": "<?xml version='1.0'?>\n<stream>\n<prompt id=\"15\">\n<text>Signature verification failed for file &apos;repomd.xml&apos; from repository &apos;google-chrome&apos;.\nWarning: This might be caused by a malicious change in the file!\nContinuing might be risky. Continue anyway?</text>\n<option value=\"yes\" desc=\"\"/>\n<option default=\"1\" value=\"no\" desc=\"\"/>\n</prompt>\n<message type=\"error\">Repository &apos;google-chrome&apos; is invalid.\n[google-chrome|http://dl.google.com/linux/chrome/rpm/stable/x86_64] Valid metadata not found at specified URL\nPlease check if the URIs defined for this repository are pointing to a valid repository.\n</message>\n<message type=\"warning\">Skipping repository &apos;google-chrome&apos; because of the above error.</message>\n<message type=\"error\">Some of the repositories have not been refreshed because of an error.</message>\n\n<search-result version=\"0.0\">\n<solvable-list>\n<solvable status=\"installed\" name=\"openSUSE-2016-106\" kind=\"patch\" edition=\"1\" arch=\"noarch\" repository=\"update-oss\"/>\n<solvable status=\"installed\" name=\"openSUSE-2016-38\" kind=\"patch\" edition=\"1\" arch=\"noarch\" repository=\"update-oss\"/>\n</solvable-list>\n</search-result>\n</stream>\n", "stdout_lines": ["<?xml version='1.0'?>", "<stream>", "<prompt id=\"15\">", "<text>Signature verification failed for file &apos;repomd.xml&apos; from repository &apos;google-chrome&apos;.", "Warning: This might be caused by a malicious change in the file!", "Continuing might be risky. Continue anyway?</text>", "<option value=\"yes\" desc=\"\"/>", "<option default=\"1\" value=\"no\" desc=\"\"/>", "</prompt>", "<message type=\"error\">Repository &apos;google-chrome&apos; is invalid.", "[google-chrome|http://dl.google.com/linux/chrome/rpm/stable/x86_64] Valid metadata not found at specified URL", "Please check if the URIs defined for this repository are pointing to a valid repository.", "</message>", "<message type=\"warning\">Skipping repository &apos;google-chrome&apos; because of the above error.</message>", "<message type=\"error\">Some of the repositories have not been refreshed because of an error.</message>", "", "<search-result version=\"0.0\">", "<solvable-list>", "<solvable status=\"installed\" name=\"openSUSE-2016-106\" kind=\"patch\" edition=\"1\" arch=\"noarch\" repository=\"update-oss\"/>", "<solvable status=\"installed\" name=\"openSUSE-2016-38\" kind=\"patch\" edition=\"1\" arch=\"noarch\" repository=\"update-oss\"/>", "</solvable-list>", "</search-result>", "</stream>"]}
```

The actual error,

```
# zypper --quiet -x --non-interactive search --match-exact -t patch openSUSE-2016-38
<?xml version='1.0'?>
<stream>
<prompt id="15">
<text>Signature verification failed for file &apos;repomd.xml&apos; from repository &apos;google-chrome&apos;.
Warning: This might be caused by a malicious change in the file!
Continuing might be risky. Continue anyway?</text>
<option value="yes" desc=""/>
<option default="1" value="no" desc=""/>
</prompt>
<message type="error">Repository &apos;google-chrome&apos; is invalid.
[google-chrome|http://dl.google.com/linux/chrome/rpm/stable/x86_64] Valid metadata not found at specified URL
Please check if the URIs defined for this repository are pointing to a valid repository.
</message>
<message type="warning">Skipping repository &apos;google-chrome&apos; because of the above error.</message>
<message type="error">Some of the repositories have not been refreshed because of an error.</message>

<search-result version="0.0">
<solvable-list>
<solvable status="installed" name="openSUSE-2016-38" summary="Security update for openssh" kind="patch"/>
</solvable-list>
</search-result>
</stream>
# echo $?
106
```

**Comment by jasonmader** [comment]
Another failure,

```
- name: 'zypper joe, vim, tkdiff, dos2unix, encfs, keepassx'
  zypper3: state=latest name=joe,vim,vim-data,tkdiff,dos2unix,sshfs,encfs,keepassx
```

```
TASK [leap : zypper joe, vim, tkdiff, dos2unix, encfs, keepassx] ***************
fatal: [ccsa-2115b.local]: FAILED! => {"changed": false, "failed": true, "module_stderr": "", "module_stdout": "Traceback (most recent call last):\r\n  File \"/root/.ansible/tmp/ansible-tmp-1461854510.89-145244766899279/zypper3\", line 2322, in <module>\r\n    main()\r\n  File \"/root/.ansible/tmp/ansible-tmp-1461854510.89-145244766899279/zypper3\", line 348, in main\r\n    retvals = package_present(module, name, state == 'latest')\r\n  File \"/root/.ansible/tmp/ansible-tmp-1461854510.89-145244766899279/zypper3\", line 260, in package_present\r\n    if m._diff:\r\nAttributeError: 'AnsibleModule' object has no attribute '_diff'\r\n", "msg": "MODULE FAILURE", "parsed": false}
    to retry, use: --limit @/Users/jason/net/playbook/site.retry
```

**Comment by robinro** [comment]
@jasonmader thanks for testing!
1) signature verification failed: I'm not sure how to deal with this best. You have a repo with broken signatures, which will procude errors on any zypper run. Did this work with the old zypper module?  I added a commit to ignore the 106 return code, which should be save because the broken repo is ignored by default. On the other hand one might overlook serious errors that way, so maybe we should actually fail in that case. I added a commit that ignored the 106 return code and also deals with code 103 (zypper itself updated) correctly, which will be useful for `state=latest name=*`
2) _diff missing: This looks like you are not using the ansible 2.1 codebase but something older. The updates zypper module will only be available with zypper 2.2 and up, so I don't see the need to check for the ansible version here. Which ansible version did you run this with?

**Comment by robinro** [comment]
wait, the error code for you is `4`, not `106`. `search` fails with `106`, which I find feasible to discard, but `patch` fails with `4`, which is very generic and could also be a broken rpm package or full disk or whatever. `4` also makes sense because the install could fail because of the missing repo.

Because `4` is very generic, I really don't want to ignore it here and the solution in my opinion would be for you to fix your repos.

**Comment by jasonmader** [comment]
@robinro I used it with my existing ansible 2.0.1.0. So no worries that it’s incompatible. I have SLES 11 here so Iʼll be able to do some testing against that too.

**Comment by robinro** [comment]
@jasonmader ok, so that explained the second part. What about the signature failure? Do you agree this is a valid reason for the module to fail?

**Comment by gregdek** [comment]
Thanks @robinro for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by robinro** [comment]
ready_for_review
The build failure is unrelated

**Comment by gregdek** [comment]
Thanks @robinro. To the current maintainers, @robinro please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by robinro** [comment]
From my side it's shipit.

**Comment by gregdek** [comment]
Thanks @robinro. Since you are one of the maintainers (@robinro) of this module and you commented with "shipit", we are marking this PR for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by robinro** [comment]
thanks @resmo for the merge

**Comment by mrummuka** [comment]
Hmm. I'm still seeing https://github.com/ansible/ansible-modules-extras/issues/2094 to occur with ansible 2.1.1.0 installed to Fedora host provisioning Opensuse 42.1 guest. 
Is this because fix is incorrectly marked for 2.1.0 or am I just interpreting "devel,  2.1.0" wrong?
What version of Ansible should I install to get this working? 

EDIT:
while
`
zypper install -y https://releases.hashicorp.com/vagrant/1.8.5/vagrant_1.8.5_x86_64.rpm 
`
or

```
- name: install vagrant directly from Hashicorp web pages
  shell: zypper install -y https://releases.hashicorp.com/vagrant/1.8.5/vagrant_1.8.5_x86_64.rpm 
  become: true

```

gets things done, 

```
- name: install vagrant directly from Hashicorp web pages #bug 2094
  zypper: name=https://releases.hashicorp.com/vagrant/1.8.5/vagrant_1.8.5_x86_64.rpm state=present
  become: true

```

fails miserably with

```
FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "No provider of 'vagrant' found.\n", 
```

**Comment by resmo** [comment]
it is not in 2.1.x, because it is a refactoring. This PR was merged to devel and will be in 2.2.

**Pull request merged**